[SPARK-20384][SQL] Support value class in nested schema for Dataset

mickjermsurawong-stripe opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error message, please read the guideline first:
     https://spark.apache.org/error-message-guidelines.html
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
- This PR revisits https://github.com/apache/spark/pull/22309, and [SPARK-20384](https://issues.apache.org/jira/browse/SPARK-20384) solving the original problem, but additionally will prevent backward-compat break on schema of top-level `AnyVal` value class.
- Why previous break? We currently support top-level value classes just as any other case class; field of the underlying type is present in schema. This means any dataframe SQL filtering on this expects the field name to be present. The previous PR changes this schema and would result in breaking current usage. See test `"schema for case class that is a value class"`. This PR keeps the schema.
- We actually currently support collection of value classes prior to this change, but not case class of nested value class. This means the schema of these classes shouldn't change to prevent breaking too. 
- However, what we can change, without breaking, is schema of nested value class, which will fails due to the compile problem, and thus its schema now isn't actually valid. After the change, the schema of this nested value class is now flattened
- With this PR, there's flattening only for nested value class (new), but not for top-level and collection classes (existing behavior)
- This PR revisits https://github.com/apache/spark/pull/27153 by handling tuple `Tuple2[AnyVal, AnyVal]` which is a constructor ("nested class") but is a generic type, so it should not be flattened behaving similarly to `Seq[AnyVal]`


### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
- Currently, nested value class isn't supported. This is because when the generated code treats `anyVal` class in its unwrapped form, but we encode the type to be the wrapped case class. This results in compile of generated code
For example, 
For a given `AnyVal` wrapper and its root-level class container
```
case class IntWrapper(i: Int) extends AnyVal
case class ComplexValueClassContainer(c: IntWrapper)
```
The problematic part of generated code:
```
    private InternalRow If_1(InternalRow i) {
        boolean isNull_42 = i.isNullAt(0);
        // 1) ******** The root-level case class we care
        org.apache.spark.sql.catalyst.encoders.ComplexValueClassContainer value_46 = isNull_42 ?
            null : ((org.apache.spark.sql.catalyst.encoders.ComplexValueClassContainer) i.get(0, null));
        if (isNull_42) {
            throw new NullPointerException(((java.lang.String) references[5] /* errMsg */ ));
        }
        boolean isNull_39 = true;
        // 2) ******** We specify its member to be unwrapped case class extending `AnyVal`
        org.apache.spark.sql.catalyst.encoders.IntWrapper value_43 = null;
        if (!false) {

            isNull_39 = false;
            if (!isNull_39) {
                // 3) ******** ERROR: `c()` compiled however is of type `int` and thus we see error
                value_43 = value_46.c();
            }
        }
```
We get this errror: Assignment conversion not possible from type "int" to type "org.apache.spark.sql.catalyst.encoders.IntWrapper"
```
java.util.concurrent.ExecutionException: org.codehaus.commons.compiler.CompileException: 
File 'generated.java', Line 159, Column 1: failed to compile: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 159, Column 1: Assignment conversion not possible from type "int" to type "org.apache.spark.sql.catalyst.encoders.IntWrapper"
```


From [doc](https://docs.scala-lang.org/overviews/core/value-classes.html) on value class: , Given: `class Wrapper(val underlying: Int) extends AnyVal`,
1) "The type at compile time is `Wrapper`, but at runtime, the representation is an `Int`". This implies that when our struct has a field of value class, the generated code should support the underlying type during runtime execution.
2) `Wrapper` "must be instantiated... when a value class is used as a type argument". This implies that `scala.Tuple[Wrapper, ...], Seq[Wrapper], Map[String, Wrapper], Option[Wrapper]` will still contain Wrapper as-is in during runtime instead of `Int`.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
- Yes, this will allow support for the nested value class.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
- Added unit tests to illustrate 
  - raw schema
  - projection
  - round-trip encode/decode

mickjermsurawong-stripe committed:
failing test on nested/complex value

mickjermsurawong-stripe committed:
assert current spark schema for collection of values

mickjermsurawong-stripe committed:
unwrap params passing encoder suite

clean up

mickjermsurawong-stripe committed:
fix test: inner value class unwrapped

mickjermsurawong-stripe committed:
clean up case class instantiation

mickjermsurawong-stripe committed:
add more test for nested collection

mickjermsurawong-stripe committed:
fix nits

mickjermsurawong-stripe committed:
add filter tests for explicitness in sql schema

mickjermsurawong-stripe committed:
handle tuple

mickjermsurawong-stripe committed:
add enc/decoder test

mickjermsurawong-stripe commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:
```diff
@@ -358,17 +417,17 @@ object ScalaReflection extends ScalaReflection {
+        val unwrappedParams = getConstructorUnwrappedParameters(t, isTypeTuple)
 
-        val arguments = params.zipWithIndex.map { case ((fieldName, fieldType), i) =>
+        val arguments = unwrappedParams.zipWithIndex.map { case ((fieldName, fieldType), i) =>
```
Deserializer expression

mickjermsurawong-stripe commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:
```diff
@@ -584,8 +643,8 @@ object ScalaReflection extends ScalaReflection {
-        val params = getConstructorParameters(t)
-        val fields = params.map { case (fieldName, fieldType) =>
+        val unwrappedParams = getConstructorUnwrappedParameters(t, tpe)
+        val fields = unwrappedParams.map { case (fieldName, fieldType) =>
```
Serializer expression

mickjermsurawong-stripe commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:
```diff
@@ -788,9 +847,9 @@ object ScalaReflection extends ScalaReflection {
+        val unwrappedParams = getConstructorUnwrappedParameters(t, tpe)
         Schema(StructType(
-          params.map { case (fieldName, fieldType) =>
+          unwrappedParams.map { case (fieldName, fieldType) =>
```
Get schema type

AmplabJenkins commented:
Can one of the admins verify this patch?

eejbyfeldt commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:
```diff
@@ -155,6 +155,65 @@ object ScalaReflection extends ScalaReflection {
+  private def getConstructorUnwrappedParameters(tpe: Type, isTupleType: Boolean):
+  Seq[(String, Type)] = {
+    val params = getConstructorParameters(tpe)
+    if (isTupleType) {
```
A scala `Tuple` is just a case class with generics: https://github.com/scala/scala/blob/2.13.x/src/library/scala/Tuple2.scala#L24

But since the current code looks specifically for `scala.Tuple` this means that it will not handle user defined case classes with generics. 

For example I think adding the following test in the `ExpressionEncodeSuite` case on this branch will fail:
```
case class CaseClassWithGeneric[T](generic: T, value: IntWrapper)
encodeDecodeTest(
  CaseClassWithGeneric[IntWrapper](IntWrapper(1), IntWrapper(2)),
  "case class with generic fields")
```

eejbyfeldt reviewed (commented):
I think this PR does not handle case class with generics types correctly, but this is probably fixable. But will probably require some more complexity in the `getConstructorUnwrappedParameters` function.


The old PR by @mt40 https://github.com/apache/spark/pull/22309 implemenation was such that a value class would *always* be encodeded by the underlying type. But in this PR it looks like in this PR encoded the elements like they are handled in the JVM, meaning that context is such that the `AnyVal` is represented as the underlying type the schema will just be the underlying type, while if it is boxed in the wrapper type this will also be reflected in the schema.

So with the following types 
```
case class IntWrapper(val i: Int) extends AnyVal
case class CaseClassWithGeneric[T](generic: T, value: IntWrapper)
```
the schema in the PR: https://github.com/apache/spark/pull/22309 would be (and is for what is currently in this branch)_
```
Schema(
  StructType(
    Seq(
      StructField("generic", IntegerType, false),
      StructField("value", IntegerType, false)
    )
  ),
  nullable = true)
)
```
But I guess the deseired behavior for the implemenation suggestion in this branch would give a schema
```
Schema(
  StructType(
    Seq(  
      StructField("generic", StructType(Seq(StructField("i", IntegerType, false)))),
      StructField("value", IntegerType, false)
    )
  ),
  nullable = true) 
)
```

I don't have any strong feeling which is the correct approach to take in the implemenation. This approach gives slightly simpler implementation, while the other one hides some of the ugliness/complexities of how AnyVals work on the JVM.


But it would be good to add some test cases with case classes with genereic fields so their behavior is covered in test cases.

eejbyfeldt commented:
I pushed in an initial implementation to a branch on my fork here: https://github.com/eejbyfeldt/spark/compare/master...support-value-classes-in-datasets?expand=1 I believe this implementation solves the cases where a case class has a generic member. My branch takes the approach that was in @mt40 original PR that value class are always encoded as the underlying type, since I am starting to think that is the correct way to implement this feature.

eejbyfeldt commented:
@mickjermsurawong-stripe  I created a WIP PR with my branch here: https://github.com/apache/spark/pull/33316  I think you can just take the `getConstructorParameters` from there to make this patch handle case classes and tuples correctly. Let me know if you want me to make a patch.

@cloud-fan You previously were involved in reviewing this PR: https://github.com/apache/spark/pull/22309  Could you provide some input with regards to the importance of backwards compatibility with schema changes for case classes. It does not seem like that was really dicussed when that patch was being proposed. 

Based on searches in the mailing lists and stackoverflow (and this being broken for so long) it would seem like a lot of people are not using this feature:

http://apache-spark-user-list.1001560.n3.nabble.com/template/NamlServlet.jtp?macro=search_page&node=1&query=value+class&days=0
http://apache-spark-developers-list.1001551.n3.nabble.com/template/NamlServlet.jtp?macro=search_page&node=1&query=value+class&days=0
https://stackoverflow.com/search?tab=newest&q=%5Bapache-spark%5D%20%22value%20class%22

mickjermsurawong-stripe commented:
hi @eejbyfeldt  please feel free to drive this forward. Thank you for the work. Happy if you'd like to make a patch here :) 
On backward-compat, i think one problem is the historical data that was written with `List[ValueClass]` will have wrapped schema. This has been working and we likely won't get good grasp from stackoverflow or mailing list.

Commit:
Properly handle ValueClass as type parameters

I also remove the unwrapped params naming, as I belive this is not
helpful. Replacing the value class with their underlying type is what
you have to do to get the correct type signature. Consider the following
example:
```
$ cat TestValueClass.scala
package example

case class ValueClass(value: Int) extends AnyVal
case class HasCaseClass(value1: ValueClass, value2: Int)
$ scalac TestValueClass.scala
$ javap -p example/HasCaseClass.class | grep 'public example.Has'
  public example.HasCaseClass copy(int, int);
  public example.HasCaseClass(int, int);
```
The Constructor for `HasCaseClass` does take two `int` and does not
mention `ValueClass`.

eejbyfeldt commented:
> hi @eejbyfeldt please feel free to drive this forward. Thank you for the work. Happy if you'd like to make a patch here :)
> On backward-compat, i think one problem is the historical data that was written with `List[ValueClass]` will have wrapped schema. This has been working and we likely won't get good grasp from stackoverflow or mailing list.

I pushed a commit here (Since I think I am not allow to push to your branch): https://github.com/eejbyfeldt/spark/commit/cbbe05e0c1c2b942e6337e62a03991ee7fa95d26 it should be based on this branch therefore you can just cherry-pick it to this branch. 

Yeah, thinking about it I agree that it better to be backwards compatible. Then this change can be seen as a bug fix for "nested value classes" and will not break an existing code, which should be a much "safer" change to do.

mickjermsurawong-stripe commented:
thanks @eejbyfeldt this is much neater!

eejbyfeldt reviewed (approved):
To me this implementation looks good! Would be great if some admin had look.

srowen commented:
Jenkins test this please

srowen commented:
Is there any behavior change you can think of that might affect users?

SparkQA commented:
**[Test build #141957 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141957/testReport)** for PR 33205 at commit [`23d0e94`](https://github.com/apache/spark/commit/23d0e94d661a5ad6898a93bbee78ca785061e754).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46468/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46468/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46468/

SparkQA commented:
**[Test build #141957 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141957/testReport)** for PR 33205 at commit [`23d0e94`](https://github.com/apache/spark/commit/23d0e94d661a5ad6898a93bbee78ca785061e754).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141957/

eejbyfeldt commented:
> Is there any behavior change you can think of that might affect users?

Hi Sean, thanks for having a look!

This only changes is for case class containing value class. e.g
```
case class IntWrapper(value: Int) extends AnyVal
case class DatasetModel(wrappedInt: IntWrapper)
```
Before this patch trying to create a `Dataset` using the `DatasetModel` would result in rumtime error like:
```
21/08/03 07:50:01 ERROR CodeGenerator: failed to compile: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 85, Column 1: Assignment conversion not possible from type "int" to type "example.IntWrapper"
org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 85, Column 1: Assignment conversion not possible from type "int" to type "example.IntWrapper"
	at org.codehaus.janino.UnitCompiler.compileError(UnitCompiler.java:12021)
	at org.codehaus.janino.UnitCompiler.assignmentConversion(UnitCompiler.java:10851)
	...
```

But with this patch it will work like expected. Unless someone explicitly depend on having this failure I don't think there should be any behavior change that is noticeable for users.

mickjermsurawong-stripe commented:
Hi Sean, I can also confirm that the impact on users here is rather fixing the currently broken behavior. There's no schema change to the previously working edge case eg. `List[AnyValClass]`

srowen commented:
Just to triple check my understanding, the following isn't a problem because it doesn't work at all now, or something else? 

"However, what we can change, without breaking, is schema of nested value class, which will fails due to the compile problem, and thus its schema now isn't actually valid. After the change, the schema of this nested value class is now flattened"

mickjermsurawong-stripe commented:
That's correct @srowen. Nested AnyVal class does not work currently. 
Value class in nested schema 1) currently does not work because the schema described has AnyVal class 2) but when accessing that nested value actually has unwrapped type `int` 3), resulting in this exception 4). Essentially, we currently describe schema in an incompatible way with how AnyVal class operates "The type at compile time is Wrapper, but at runtime, the representation is an Int". ([doc](https://docs.scala-lang.org/overviews/core/value-classes.html))
```
    private InternalRow If_1(InternalRow i) {
        boolean isNull_42 = i.isNullAt(0);

########################## 1) The root-level case class we care ##########################

        org.apache.spark.sql.catalyst.encoders.ComplexValueClassContainer value_46 = isNull_42 ?
            null : ((org.apache.spark.sql.catalyst.encoders.ComplexValueClassContainer) i.get(0, null));
        if (isNull_42) {
            throw new NullPointerException(((java.lang.String) references[5] /* errMsg */ ));
        }
        boolean isNull_39 = true;

########################## 2) We specify its member to be unwrapped case class extending `AnyVal`

        org.apache.spark.sql.catalyst.encoders.IntWrapper value_43 = null;
        if (!false) {

            isNull_39 = false;
            if (!isNull_39) {

########################## 3) ******** ERROR: `c()` compiled however is of type `int` and thus we see error

                value_43 = value_46.c();
            }
        }
```
4) 
```
java.util.concurrent.ExecutionException: org.codehaus.commons.compiler.CompileException: 
File 'generated.java', Line 159, Column 1: failed to compile: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 159, Column 1: Assignment conversion not possible from type "int" to type "org.apache.spark.sql.catalyst.encoders.IntWrapper"
```

To your specific clarification "because it doesn't work at all now": it _does_ work in one case of value class in parameterized class like `Seq[AnyVal]`. This is because there is no unwrapping, and the wrapper remains as-is. From the same scala doc [ref](https://docs.scala-lang.org/overviews/core/value-classes.html), `Wrapper` "must be instantiated... when a value class is used as a type argument". This implies that `scala.Tuple[Wrapper, ...], Seq[Wrapper], Map[String, Wrapper], Option[Wrapper]` will still contain Wrapper as-is in during runtime instead of `Int`.

This fix will also resolve schema issue [SPARK-20384 ](https://issues.apache.org/jira/browse/SPARK-20384) originally described; the reporter will be able to access the value class in an unwrapped fashion.

srowen commented:
Merged to master

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/ScalaReflectionSuite.scala:
```diff
@@ -451,4 +462,87 @@ class ScalaReflectionSuite extends SparkFunSuite {
+  }
+
+  test("SPARK-20384: schema for tuple_2 of value class") {
+    val schema = schemaFor[(IntWrapper, StrWrapper)]
```
It's a bit weird that the schema of case class of value classes is not consistent with the schema of tuple of value classes, but there seems no better solution as we need to keep backward compatibility.
